src/main/java/frc/utils/auto/PathPlannerUtils.java

```diff
@@ -84,6 +89,57 @@ public static void registerCommand(String commandName, Command command) {
 		NamedCommands.registerCommand(commandName, command);
 	}
 
+	public static Pose2d getAllianceRelativePose(Pose2d bluePose) {
+		if (AutoBuilder.shouldFlip()) {
+			return FlippingUtil.flipFieldPose(bluePose);
+		}
+		return bluePose;
+	}
+
+	public static Optional<PathPlannerPath> getPathFromFile(String pathName, String logPath) {
+		try {
+			return Optional.of(PathPlannerPath.fromPathFile(pathName));
+		} catch (Exception exception) {
+			new Alert(Alert.AlertType.ERROR, logPath + exception.getMessage()).report();
+		}
+		return Optional.empty();
+	}
+
+	public static Pose2d getPathStartingPose(PathPlannerPath path) {
+		return new Pose2d(path.getPathPoses().get(0).getTranslation(), path.getIdealStartingState().rotation());
+	}
+
+	public static Command followPath(PathPlannerPath path) {
+		return AutoBuilder.followPath(path);
+	}
+
+	public static Command pathfindToPose(Pose2d targetPose, PathConstraints pathfindingConstraints) {
+		return AutoBuilder.pathfindToPose(targetPose, pathfindingConstraints);
+	}
+
+	public static Command pathfindThenFollowPath(PathPlannerPath path, PathConstraints pathfindingConstraints) {
+		return AutoBuilder.pathfindThenFollowPath(path, pathfindingConstraints);
+	}
+
+	public static boolean isRobotCloseToPathBeginning(PathPlannerPath path, Supplier<Pose2d> currentPose, double toleranceMeters) {
+		return getAllianceRelativePose(path.getPathPoses().get(0)).getTranslation().getDistance(currentPose.get().getTranslation())
+			<= toleranceMeters;
+	}
+
+	public static Pose2d getLastPathPose(PathPlannerPath path) {
+		return new Pose2d(path.getPathPoses().get(path.getPathPoses().size() - 1).getTranslation(), path.getGoalEndState().rotation());
+	}
+
+	public static Command followPathOrDriveToPathEnd(Robot robot, PathPlannerPath path) {
+		return robot.getSwerve()
+			.getCommandsBuilder()
+			.followPathOrDriveToPathEnd(
+				robot.getPoseEstimator()::getCurrentPose,
+				path,
+				AutonomousConstants.CLOSE_TO_TARGET_POSITION_DEADBAND_METERS
+			);
+	}
```

Comment:
i dont think its should be here (subsystem usage in utils)

Reply:
so where to put it? or do you want me to duplicate the code each time I follow path?

Reply:
U can create in robot on auto package class specific for command of pp that use robot and keep in utils just utils.

U can put in swerve and get suppliers.